ddtrace/propagation/utils.py

```diff
@@ -6,17 +6,40 @@ def get_wsgi_header(header):
     return "HTTP_{}".format(header.upper().replace("-", "_"))
 
 
+_CACHE_HEADER_MAX_SIZE = 256
+_cached_headers = {}
+
+
+def _shrink_cache():
+    for _, h in zip(range(_CACHE_HEADER_MAX_SIZE >> 1), sorted(_cached_headers, key=lambda h: _cached_headers[h][1])):
+        del _cached_headers[h]
```

Comment:
Why not using something more standard like `functools.lru_cache`?

Reply:
A couple of reasons:
- `lru_cache` is not available in Python 2
- LRU is perhaps not the best strategy as some headers will be more _frequent_ than others, so MFU seemed more suited to the problem.
- The function we are caching is very simple and a generic caching utility like `lru_cache` would add quite some overhead (e.g. by calling stuff to try and serialise `kwargs` ...)

Reply:
Right, but I'm really not excited by having custom caching code spread everywhere over the codebase. :)
Could we leverage something like https://pypi.org/project/cachetools/ even if we have to vendorize it?

Reply:
We could, but by looking at the code (https://github.com/tkem/cachetools/blob/master/cachetools/decorators.py) it looks like the overhead would be comparable to `lru_cache` (the `@cached` decorator adds a wrapper function in between, which then has to call a `key` function for no particular reason other than returning the sole argument passed to it, plus a few levels of branching). I also don't quite like putting ad-hoc caching logic in random places, but given that the whole point of the exercise is performance perhaps we can make an exception in this case 🙂

Reply:
I think we should evaluate:

1. What is the impact and improvement running a realistic web application workload (hello world is an example, but exaggerates our overheads, something that actually hits a db/does some processing is usually more realistic)
2. Can we avoid calls to these functions all together? It seems like we can, in which case, if we can reduce the total number of calls by X% then it probably won't matter if we have a cache here or not (improve performance, but keep the code simple)